### ISSUE TITLE
Panic: CARGO_MANIFEST_DIR must not change within one compiler process

### ISSUE THREAD

**Issue by djc** [rust-lang/rust-analyzer#12969]
**rust-analyzer version**: rust-analyzer version: 0.3.1162-standalone (634cfe3d7 2022-08-07)

**rustc version**: 1.62.1 (e092d0b6b 2022-07-16)

**relevant settings**: VS Code, fairly vanilla setup (except that my VS Code workspace is rooted in a directory that contains multiple Cargo workspaces, perhaps?). Both of the crates mentioned below are in the same Cargo workspace though, so I suppose that shouldn't be an issue?

```
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `"/Users/djc/src/bolt/rust/domains/epoxide"`,
 right: `"/Users/djc/src/bolt/rust/domains/store"`: CARGO_MANIFEST_DIR must not change within one compiler process', /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/proc-macro-crate-1.2.0/src/lib.rs:137:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panicking.rs:142:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panicking.rs:181:5
   4: proc_macro_crate::crate_name
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/proc-macro-crate-1.2.0/src/lib.rs:137:5
   5: async_graphql_derive::utils::get_crate_name
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-derive-4.0.6/src/utils.rs:40:26
   6: async_graphql_derive::simple_object::generate
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-derive-4.0.6/src/simple_object.rs:29:22
   7: async_graphql_derive::derive_simple_object
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-derive-4.0.6/src/lib.rs:52:11
   8: core::ops::function::FnOnce::call_once
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/ops/function.rs:248:5
   9: proc_macro::bridge::client::Client<fn(proc_macro::TokenStream) .> proc_macro::TokenStream>::expand1::run::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:424:40
  10: proc_macro::bridge::client::run_client::{{closure}}::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:392:26
  11: proc_macro::bridge::scoped_cell::ScopedCell<T>::set::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/scoped_cell.rs:79:33
  12: proc_macro::bridge::scoped_cell::ScopedCell<T>::replace
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/scoped_cell.rs:74:9
  13: proc_macro::bridge::scoped_cell::ScopedCell<T>::set
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/scoped_cell.rs:79:9
  14: proc_macro::bridge::client::<impl proc_macro::bridge::Bridge>::enter::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:340:35
  15: std::thread::local::LocalKey<T>::try_with
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/thread/local.rs:445:16
  16: std::thread::local::LocalKey<T>::with
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/thread/local.rs:421:9
  17: proc_macro::bridge::client::<impl proc_macro::bridge::Bridge>::enter
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:340:9
  18: proc_macro::bridge::client::run_client::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:385:9
  19: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panic/unwind_safe.rs:271:9
  20: std::panicking::try::do_call
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:492:40
  21: ___rust_try
  22: std::panicking::try
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:456:19
  23: std::panic::catch_unwind
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panic.rs:137:14
  24: proc_macro::bridge::client::run_client
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:384:5
  25: proc_macro::bridge::client::Client<fn(proc_macro::TokenStream) .> proc_macro::TokenStream>::expand1::run
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:424:13
  26: proc_macro_srv::abis::abi_1_58::proc_macro::bridge::server::run_server
  27: proc_macro_srv::abis::abi_1_58::Abi::expand
  28: proc_macro_srv::dylib::Expander::expand
  29: core::ops::function::FnOnce::call_once{{vtable.shim}}
```

cc @fasterthanlime I suppose you might know something about this?

**Comment by Veykril** [comment]
The description of `CARGO_MANIFEST_DIR` is `The directory containing the manifest of your package.`. So isn't this behavior correct? That is the env var points to the manifest of the crate, not the workspace.

**Comment by djc** [comment]
Yes, the env var points to the manifest of the crate -- that is correct, and I'm not objecting to that. But it seems that somehow that Rust Analyzer is somehow reusing some process/environment in a way it shouldn't?

Ah, I notice that this is happening in the proc_macro_crate crate, so maybe it is relying on an invariant that doesn't actually exist, or that has been upheld so far in the rustc implementation of the proc-macro setup but not in the new RA setup?

**Comment by lnicola** [comment]
So we start the proc macro server once, load the relevant macro libs and expand each invocation in a new thread, to allow the bridge implementation to use thread-local variables.

I think that `OnceCell` gets initialized once, then reused when compiling another crate, because we don't unload the libraries.

**Comment by djc** [comment]
Yup. I've opened an issue against proc-macro-crate, too: https://github.com/bkchr/proc-macro-crate/issues/22. Not sure who's "right" here?

**Comment by lnicola** [comment]
> that has been upheld so far in the rustc implementation of the proc-macro setup but not in the new RA setup

It's not related to RA, it's a recent change in that crate: https://github.com/bkchr/proc-macro-crate/commit/47c2732f422fcba0730de3baf9823dcf8cd24ae7.

I suppose it's a bug on our side as long as we're not compatible with `rustc`, but on the other hand, I wish proc macros would rely less, not more, on their ambient environment.

**Comment by fasterthanlime** [comment]
I wouldn't necessarily close this: it's not necessarily a hard fix on the RA side: we know the crate where we the macro we're expanding has been invoked from, so we could definitely set the environment variable there.

Unless there's reluctance to spawn even more proc macro servers processes? Come to think of it, with the caching going on, I'm not even sure we correctly pick up on changes to proc macro crates (code changes if they're in the cargo workspace, or Cargo manifest changes if they're third party dependencies)

There's probably a couple more tests to add there, and a couple issues to resolve.

**Comment by lnicola** [comment]
I think we still want to keep track of this issue. We could start one proc macro server per crate, or try to fill in the environment in a more complete way, as @fasterthanlime suggested above.

Indeed, we never unload the macros, which is actually a bigger problem.

**Comment by Veykril** [comment]
> I suppose it's a bug on our side as long as we're not compatible with rustc, but on the other hand, I wish proc macros would rely less, not more, on their ambient environment.

It's not really a bug on our side though. Technically this is not something they should be doing in the first place and its "sheer luck" that it works with rustc right now because proc-macros arent really sandboxed. So I would be inclined to say this is a wont-fix if possible

**Comment by fasterthanlime** [comment]
> It's not really a bug on our side though. Technically this is not something they should be doing in the first place and its "sheer luck" that it works with rustc right now because proc-macros arent really sandboxed.

I think we should look for an authoritative answer on this. cargo is pretty clear what environment variables it sets for build scripts, for example: I'm not sure what the actual rules are for proc macros. If proc macros are never supposed to read env vars, then it's not an RA bug. If cargo actually guarantees some env vars are present for proc macros, I think RA should do its best to reproduce that as closely as possible.

> Indeed, we never unload the macros, which is actually a bigger problem, but seems much harder to fix.

This probably warrants a separate issue

**Comment by bjorn3** [comment]
Proc macros are allowed to read env vars I think. (The read should go through `proc_macro::tracked_env::var` though once it is stabilized to ensure cargo will rebuild if the env var changes.) What they are not allowed to do is keep any state across macro invocations. The problem in this case is that a proc macro cached a value derived from an env var and expected it to still be the same on the next invocation.

**Comment by Veykril** [comment]
> I think we should look for an authoritative answer on this. cargo is pretty clear what environment variables it sets for build scripts, for example: I'm not sure what the actual rules are for proc macros. If proc macros are never supposed to read env vars, then it's not an RA bug. If cargo actually guarantees some env vars are present for proc macros, I think RA should do its best to reproduce that as closely as possible.

That is actually not relevant to this issue as we are properly setting the cargo env vars here. The proc-macro is assuming that the globals it uses are per crate though, which works for rustc but not r-a, and to my knowledge the general stance regarding this is that the behavior rustc "exposes" for proc-macros here right now is not definitive.

**Comment by fasterthanlime** [comment]
> That is actually not relevant to this issue as we are properly setting the cargo env vars here.

Ahhh, so what the proc-macro crate should do is either 1) not use a cache at all, or 2) key their cache entries on `CARGO_MANIFEST_DIR` instead of asserting it doesn't change (which isn't guaranteed by anything and just happens to be cargo+rustc's behavior right now.) I see.

**Comment by jplatte** [comment]
Yeah, I failed to sufficiently think about how rust-analyzer handles proc-macros. That assertion was not based on any documented guarantees, just on "rustc is one process per crate compilation, so this has to be fine for the foreseeable future".

Luckily I decided that there had to be an assertion when relying on non-documented behavior (partially because a bug from a mismatch between the disk contents and cache would be horrifying). So that assertion perfectly fulfilled its task 😅
